### PR TITLE
Update to hotshot rc-0.5.67 / marketplace

### PR DESCRIPTION
**Pull request by tbro**
Update to hotshot rc-0.5.66

Most changes are being brought in from #1696. Also includes changes
from #1781. Most compile errors are fixed. Depends on
`hotshot-query-service` pointing to the same hotshot tab (see
Cargo.toml patch).

### CODE REVIEW

File: types/src/v0/v0_1/state.rs [hunk 1]
```diff
@@ -1,14 +1,12 @@
-use crate::{Header, ResolvableChainConfig};
-
-use super::{FeeAccount, FeeAmount};
-
+use super::{FeeAccount, FeeAmount, Header};
```
Comment (imabdulbasit) [inline types/src/v0/v0_1/state.rs:1]: oh the issue is here. It is importing v01 Header instead of that versioned header

**Review by shenkeyao — APPROVED** [review]

File: types/src/v0/impls/state.rs [hunk 2]
```diff
@@ -283,32 +303,41 @@ impl From<MerkleTreeError> for FeeError {
...
+        state.charge_fee(fee_info, recipient)?;
+        delta.fees_delta.extend([fee_info.account, recipient]);
+    }
     Ok(())
 }
 
-/// Validate builder account by verifying signature
+/// Validate builder accounts by verifying signatures. All fees are
+/// verified against signature by index.
 fn validate_builder_fee(proposed_header: &Header) -> Result<(), BuilderValidationError> {
```
Comment (QuentinI) [inline types/src/v0/impls/state.rs:318]: This should be versioned instead of replaced, pre-marketplace we use old behavior.

Comment (tbro) [reply types/src/v0/impls/state.rs:318]: *I think* this is okay.  To support old behavior, Header v1 logic has been updated to wrap fees and signatures in a `Vec`. So I think this should be ok *in principal*. I'll go over it again to look for implementation errors  though.

Comment (QuentinI) [reply types/src/v0/impls/state.rs:318]: Oh, I see. I'm seeing `invalid builder fee: Invalid Builder Signature` errors in `demo-native` though so something somewhere doesn't work.

Comment (tbro) [reply types/src/v0/impls/state.rs:318]: We were definitely missing a test for this. I've add a test and it is working as expected. Is it possible that signing is being done incorrectly? 4e3b484546acf27888e9699bdfdeb9a7822b67d3

Comment (tbro) [reply types/src/v0/impls/state.rs:318]: Sorry I wasn't thinking clearly.  c4a9a8edff1ccfff2ef912e4b0d0b53a13ea778a should fix it though.

File: types/src/v0/v0_3/chain_config.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    fn from(
+        &v0_1::ResolvableChainConfig { chain_config }: &v0_1::ResolvableChainConfig,
+    ) -> ResolvableChainConfig {
+        match chain_config {
+            Either::Left(chain_config) => ResolvableChainConfig {
+                chain_config: Either::Left(ChainConfig::from(chain_config)),
+            },
+            // TODO does this work? is there a better way?
+            Either::Right(c) => ResolvableChainConfig {
+                chain_config: Either::Right(Commitment::from_str(&c.to_string()).unwrap()),
```
Comment (imabdulbasit) [inline types/src/v0/v0_3/chain_config.rs:115]: I also haven't found a better way to handle this. maybe @jbearer would know

File: types/src/v0/v0_3/chain_config.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            fee_recipient,
+        }
+    }
+}
+
+impl Default for ChainConfig {
+    fn default() -> Self {
+        Self {
+            chain_id: U256::from(35353).into(), // arbitrarily chosen chain ID
+            max_block_size: 10240.into(),
```
Comment (imabdulbasit) [inline types/src/v0/v0_3/chain_config.rs:168]: dev node test will fail. I had to change it to 30mb

File: types/src/v0/impls/auction.rs [hunk 5]
```diff
@@ -213,19 +230,61 @@ impl BidTx {
...
-        vec![self.url.clone()]
+        self.winning_bids()
+            .iter()
+            .map(|bid| bid.url())
+            .chain(self.reserve_bids().iter().map(|bid| bid.1.clone()))
+            .collect()
+    }
+}
+
+const SOLVER_URL: &str = "https://solver:1234";
```
Comment (imabdulbasit) [inline types/src/v0/impls/auction.rs:269]: this should come from env var?

File: sequencer/Cargo.toml [hunk 6]
```diff
@@ -11,6 +11,9 @@ testing = ["hotshot-testing"]
 libp2p = []
 benchmarking = []
 
+[lints.rust]
```
Comment (imabdulbasit) [inline sequencer/Cargo.toml:14]: I think these were already by jeb and merged in main

File: data/genesis/demo.toml [hunk 7]
```diff
@@ -22,3 +22,10 @@ base_fee = '1 wei'
 max_block_size = '1mb'
 fee_recipient = '0x0000000000000000000000000000000000000000'
 fee_contract = '0xa15bb66138824a1c7167f5e85b957d04dd34e468'
+
+[[upgrade]]
+version = "0.4"
+start_proposing_view = 5
+stop_proposing_view = 15
+
+[upgrade.marketplace]
```
Comment (imabdulbasit) [inline data/genesis/demo.toml]: we can remove it for now as UpgradeType changes are in integrations PR

File: types/src/v0/impls/auction.rs (hunk 5 above)
Comment (tbro) [reply types/src/v0/impls/auction.rs:269]: Yes but it doesn't matter yet.

File: sequencer/Cargo.toml (hunk 6 above)
Comment (imabdulbasit) [reply sequencer/Cargo.toml:14]: nvm they are moved from bottom to top

File: types/src/v0/v0_3/chain_config.rs (hunk 3 above)
Comment (tbro) [reply types/src/v0/v0_3/chain_config.rs:115]: I think this is ok to merge for now, but we try to figure out if there is a better way before we turn on marketplace.

File: types/src/v0/impls/auction.rs [hunk 8]
```diff
@@ -137,6 +150,9 @@ pub enum ExecutionError {
     #[error("Could not resolve `ChainConfig`")]
     /// Could not resolve `ChainConfig`.
     UnresolvableChainConfig,
+    #[error("Bid ricipient not set on `ChainConfig`")]
```
Comment (imabdulbasit) [inline types/src/v0/impls/auction.rs:153]: typo here: Bid recipient

**Review by imabdulbasit — APPROVED** [review]
LGTM! left some minor comments.

File: data/genesis/demo.toml (hunk 7 above)
Comment (imabdulbasit) [reply data/genesis/demo.toml]: I think that toml parsing will fail so yeah needs to be removed

**Pull request merged**